### PR TITLE
Eric s geo coder

### PR DESCRIPTION
**Pull request by rraub**

### CODE REVIEW

File: test/spec/SearchApi/Support/GeoParserSpec.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * GeoParserSpec - Spec integration test for the Geo Parser Geocoder (higher level functions)
+ */
+class GeoParserSpec extends ObjectBehavior {
+  function it_is_initializable() {
+    $this->shouldHaveType( 'SearchApi\Support\GeoParser' );
+  }
+
+  function it_should_return_a_result() {
+  	$this->reverse_geocoder_parser( 'test' )->shouldBeString();
```
Comment (rraub) [inline test/spec/SearchApi/Support/GeoParserSpec.php]: Your indentation isn't consistent here, I'll have to look into why the coding standards didn't catch this.

File: src/providers/google-reverse-geocoder.php [hunk 2]
```diff
@@ -2,11 +2,30 @@
...
  * location via GoogleMaps.
  */
 class GoogleReverseGeocoder implements ReverseGeocoder {
+  public $api_key;
+
+  public function __construct( $key = null ) {
+  	$this->api_key = $key;
+  }
+
+  public function get_locations( Models\GeoCoordinate $geo_coordinate ) {
```
Comment (rraub) [inline src/providers/google-reverse-geocoder.php:29]: Could you add some basic comments to this function?

File: src/providers/google-reverse-geocoder.php [hunk 3]
```diff
@@ -2,11 +2,30 @@
...
+  	$parser = new Support\GeoParser();
+  	$mock = new Mocks\Google_Geocoder_Mock();
+  	$geocoding_results = $mock->reverse_geocoding( $geo_coordinate, $this->api_key );
+  	if ( $geocoding_results === 'Invalid Key' ) {
+  	  return 'Invalid Key';
```
Comment (rraub) [inline src/providers/google-reverse-geocoder.php]: Returning a string as an error state might not be the easiest thing for who ever is using this to catch. Especially when it looks like your actual data being returned is also be a string.

Comment (idmontie) [inline src/providers/google-reverse-geocoder.php]: It would be better to just throw an Exception with the message "Invalid Key"; because, well, its an exceptional circumstance.

File: test/spec/SearchApi/Support/GeoParserSpec.php (hunk 1 above)
Comment (ESmith92) [inline test/spec/SearchApi/Support/GeoParserSpec.php]: I checked my code and it appears to have the correct indenting in code but git shows it over farther

Comment (idmontie) [inline test/spec/SearchApi/Support/GeoParserSpec.php]: Is it being committed with an actual tab? You should set your IDE  to convert tabs to spaces, not just have them show up as 2 spaces.

Comment (ESmith92) [inline test/spec/SearchApi/Support/GeoParserSpec.php]: it should be set to 2 spaces, ill double check though

**Comment by ESmith92** [comment]
alright this branch is ready to be checked for a merge

File: src/support/geocoder-results-parser.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+
+use Exception;
+use SearchApi\Models as Models;
+
+/**
+ * Class Geo_Parser - Parses the returned result from a Geocoder
+ */
+class GeoParser {
+  public function reverse_geocoder_parser( $json_results ) {
```
Comment (rraub) [inline src/support/geocoder-results-parser.php]: This function is doing too much, you can slim it up by separating out the creation of the $search_term_array into another function and have reverse_geocoder_parser call that. This will make testing easier too since you can test the logic for creating the $search_term_array separately from the logic that encodes and decodes the json response.

File: test/spec/SearchApi/Support/GeoParserSpec.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * GeoParserSpec - Spec integration test for the Geo Parser Geocoder (higher level functions)
+ */
+class GeoParserSpec extends ObjectBehavior {
+  function it_is_initializable() {
+    $this->shouldHaveType( 'SearchApi\Support\GeoParser' );
+  }
+
+  function it_should_return_an_empty_array() {
+  	$this->reverse_geocoder_parser( '{"test":"test"}' )->shouldBeArray();
```
Comment (rraub) [inline test/spec/SearchApi/Support/GeoParserSpec.php]: This doesn't actually test if the result is an empty array.

File: test/spec/SearchApi/Providers/GoogleReverseGeocoderSpec.php [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+/**
+ * GoogleReverseGeocoderSpec - Spec integration test for the Google Reverse Geocoder (higher level functions)
+ */
+class GoogleReverseGeocoderSpec extends ObjectBehavior {
+  function it_is_initializable() {
+    $this->shouldHaveType( 'SearchApi\Providers\GoogleReverseGeocoder' );
+  }
+
+  function it_should_have_key_null() {
```
Comment (rraub) [inline test/spec/SearchApi/Providers/GoogleReverseGeocoderSpec.php]: maybe a better name is `it_shouldnt_have_an_api_key_by_default`

File: src/support/geocoder-results-parser.php (hunk 4 above)
Comment (ESmith92) [inline src/support/geocoder-results-parser.php]: Alright, I'll try to do this tonight

File: src/providers/google-reverse-geocoder.php [hunk 7]
```diff
@@ -2,11 +2,58 @@
...
  * location via GoogleMaps.
  */
 class GoogleReverseGeocoder implements ReverseGeocoder {
+  public $api_key; // service key
+  private $geo_parser; // injected parser
+
+  public function __construct( $key = null, $parser = null ) {
+  	$this->api_key = $key;
+
+  	if ( $parser === null ) {
```
Comment (rraub) [inline src/providers/google-reverse-geocoder.php]: The indentation is off here too, it has a mix of tabs and spaces

File: test/spec/SearchApi/Providers/GoogleReverseGeocoderSpec.php (hunk 6 above)
Comment (ESmith92) [inline test/spec/SearchApi/Providers/GoogleReverseGeocoderSpec.php]: It was a test to make sure it is null by default.
How about it_should_have_null_for_api_key

File: src/providers/google-reverse-geocoder.php (hunk 7 above)
Comment (ESmith92) [inline src/providers/google-reverse-geocoder.php]: Ill check if there a way to tell the ide i am using to use space when I hit enter... I think that is the issue happening with these tabs.

File: test/spec/SearchApi/Providers/GoogleReverseGeocoderSpec.php (hunk 6 above)
Comment (rraub) [inline test/spec/SearchApi/Providers/GoogleReverseGeocoderSpec.php]: Your testing the default case, which is why I would like to see the word 'default' in the test name. Otherwise it might sound more like you are just testing how the class works with a null api_key.

Comment (ESmith92) [inline test/spec/SearchApi/Providers/GoogleReverseGeocoderSpec.php]: Ah, I see what you mean

Comment (ESmith92) [inline test/spec/SearchApi/Providers/GoogleReverseGeocoderSpec.php]: Oh I just reread your suggested name, I missed the "shouldn't" and read it as "should"

**Comment by ESmith92** [comment]
this is ready to be check again

**Comment by idmontie** [comment]
:+1: Looks good. I'll let @rraub check it in, but he is out today (Wednesday).

File: test/spec/SearchApi/Support/GeoParserSpec.php [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+  	$result->shouldBeArray();
+  	$result->shouldHaveCount( 0 );
+  }
+
+  function it_should_return_an_empty_array_when_passed_bad_formatted_array() {
+  	$result = $this->reverse_geocoder_parser( array( 'a', 'b', 'f' ) );
+  	$result->shouldBeArray();
+  	$result->shouldHaveCount( 0 );
+  }
+  // end of reverse_geocoder_parser tests
```
Comment (rraub) [inline test/spec/SearchApi/Support/GeoParserSpec.php:111]: Where is the test case for the reverse_geocoder_parser returning a real value? Not an empty array.

**Comment by rraub** [comment]
:+1:
